src/vos/vos_internal.h

```diff
@@ -1289,31 +1289,52 @@ vos_iterate_key(struct vos_object *obj, daos_handle_t toh, vos_iter_type_t type,
 /** Start epoch of vos */
 extern daos_epoch_t	vos_start_epoch;
 
-/* Slab allocation */
-enum {
-	VOS_SLAB_OBJ_NODE	= 0,
-	VOS_SLAB_KEY_NODE	= 1,
-	VOS_SLAB_SV_NODE	= 2,
-	VOS_SLAB_EVT_NODE	= 3,
-	VOS_SLAB_EVT_DESC	= 4,
-	VOS_SLAB_OBJ_DF		= 5,
-	VOS_SLAB_EVT_NODE_SM	= 6,
-	VOS_SLAB_MAX		= 7
+/** Define common slabs.  We can refine this for 2.4 pools but that is for next patch */
```

Comment:
I think PMDK allocator has already internally defined/used common allocation classes, which is quite similar to "common slabs", the major difference of 'common slab' with the allocation classes is the granularity?

I think the point of using slab is to totally eliminate internal fragmentations when the object size is known (and the object is massively used). Imagine an object size is 33 bytes, with this "common slabs", it will still have almost 50% internal fragmentations.

Reply:
I'll have to look at how they do allocation when not using slabs.  My understanding from prior discussions with Piotr was that they have some overhead, probably to store back pointer with allocation whereas the slab allocation doesn't because everything in the block is the same size.

My plan was to do this so that I can then resize the various tree nodes to fit in 512 for better allocation locality (e.g fewer blocks overall).  The other reason for doing this is because the existing code didn't handle various sizes of tree nodes in VOS.

Anyway, let me do some digging on the allocations in PMDK.

Reply:
One more thing, I don't believe we have any situations where the overhead is 50%

Reply:
Ok, I have dug into it a little.   The default PMDK allocation classes all have 16 byte headers attached to each allocation.    They have only a handfull of classes in the size ranges we are creating in DAOS.  Anyway, suppose you want to allocation something that is 128 bytes, if I am reading the code correclty, that would fall into the class with 192 bytes because of the header.

Here are the classes defined in the range we are talking about with existing slab allocations.
class 0 unit_size=262144 alignment=0 units_per_block=0, header_type=1, class_id=0
class 1 unit_size=128 alignment=0 units_per_block=2045, header_type=1, class_id=1
class 2 unit_size=192 alignment=0 units_per_block=1364, header_type=1, class_id=2
class 3 unit_size=320 alignment=0 units_per_block=818, header_type=1, class_id=3
class 4 unit_size=448 alignment=0 units_per_block=584, header_type=1, class_id=4
class 5 unit_size=704 alignment=0 units_per_block=372, header_type=1, class_id=5
class 6 unit_size=832 alignment=0 units_per_block=315, header_type=1, class_id=6
....

So, for evtree node, which is presently 768 bytes, we would be in the 832 byte class.   In my PR, we would be in the 768 class in this case, wasting at most 31 bytes in any given allocation to internal fragmentation.

For a point of reference, these are all of the custom ones defined with this patch
class 54 unit_size=32 alignment=0 units_per_block=8158, header_type=2, class_id=54
class 55 unit_size=64 alignment=0 units_per_block=4087, header_type=2, class_id=55
class 56 unit_size=96 alignment=0 units_per_block=2726, header_type=2, class_id=56
class 57 unit_size=128 alignment=0 units_per_block=2045, header_type=2, class_id=57
class 58 unit_size=160 alignment=0 units_per_block=1636, header_type=2, class_id=58
class 59 unit_size=192 alignment=0 units_per_block=1364, header_type=2, class_id=59
class 60 unit_size=224 alignment=0 units_per_block=1169, header_type=2, class_id=60
class 61 unit_size=288 alignment=0 units_per_block=1819, header_type=2, class_id=61
class 62 unit_size=384 alignment=0 units_per_block=1364, header_type=2, class_id=62
class 63 unit_size=416 alignment=0 units_per_block=1259, header_type=2, class_id=63
class 64 unit_size=576 alignment=0 units_per_block=1365, header_type=2, class_id=64
class 65 unit_size=672 alignment=0 units_per_block=1170, header_type=2, class_id=65
class 66 unit_size=768 alignment=0 units_per_block=1023, header_type=2, class_id=66

I plan to have fewer allocation classes (that we actually use) in a subsequent patch to try to avoid allocating too many slabs and also try to avoid alignment issues due to weird sizes.

Reply:
Sounds reasonable to me, thanks for looking into PMDK allocation classes.